openhands/server/session/session.py

```diff
@@ -138,12 +139,19 @@ async def on_event(self, event: Event):
             return
         if event.source == EventSource.AGENT:
             await self.send(event_to_dict(event))
-        elif event.source == EventSource.USER and isinstance(
+        # NOTE: ipython observations are not sent here currently
+        elif event.source == EventSource.ENVIRONMENT and isinstance(
             event, CmdOutputObservation
         ):
-            await self.send(event_to_dict(event))
+            # feedback from the environment to agent actions is understood as agent events by the UI
+            event_dict = event_to_dict(event)
+            event_dict['source'] = EventSource.AGENT
+            await self.send(event_dict)
         elif isinstance(event, ErrorObservation):
-            await self.send(event_to_dict(event))
+            # send error events as agent events to the UI
+            event_dict = event_to_dict(event)
+            event_dict['source'] = EventSource.AGENT
```

Comment:
Do we have to do this - if the "source" is not *that* impactful in the FE as @amanape suggested?

Reply:
Actually, let me run this PR on my machine later today to be sure. The only place I know `source` is used are for displaying the right message variant. I'll triple check and submit my review, too

Reply:
The purpose here is to _not_ impact the UI at all. We would send exactly what it had before. If the UI wants or needs ENV in the future, we can change this code.

The wider picture here is simply a matter of separating concerns a little better: 
- the `agent session` is a subscriber to the event stream. It sends events to the UI. It will format and store in the `dict` whatever the UI needs from the event that just happened.
- the [(in)famous little method](https://github.com/All-Hands-AI/OpenHands/blob/main/openhands/controller/agent_controller.py) in the controller should not need to adapt things _for the UI's consumption_. I hope we can completely remove UI-intended code from the `agent controller` on @rbren PR. Hey, if it's not there, we won't break it in the future. One can hope! 😅